internal/transport/transport_test.go

```diff
@@ -2758,21 +2749,17 @@ func (s) TestClientSendsAGoAwayFrame(t *testing.T) {
 }
 
 // hangingConn is a net.Conn wrapper for testing, simulating hanging connections
-// after a GOAWAY frame is sent, of which Write operations pause until explicitly signaled
-// or a timeout occurs.
+// after a GOAWAY frame is sent, of which Write operations pause until explicitly
+// signaled or a timeout occurs.
 type hangingConn struct {
 	net.Conn
-	hangConn chan struct{}
-}
-
-func (hc *hangingConn) Read(b []byte) (n int, err error) {
-	n, err = hc.Conn.Read(b)
-	return n, err
+	hangConn       chan struct{}
+	isGreetingDone *atomic.Bool
 }
 
 func (hc *hangingConn) Write(b []byte) (n int, err error) {
 	n, err = hc.Conn.Write(b)
-	if isGreetingDone.Load() == true {
+	if hc.isGreetingDone.Load() == true {
 		// Hang the Write for more than goAwayLoopyWriterTimeout
```

Comment:
Remove this timer?  Otherwise it will release on its own and pass on master@HEAD, won't it?

Reply:
Ping, this doesn't seem done.

Reply:
@dfawley After having more thought on this, I think we would need to have timeout for `hc.hangConn`, otherwise [this](https://github.com/grpc/grpc-go/blob/master/internal/transport/http2_client.go#L467) go-routine will be leaked via [this](https://github.com/grpc/grpc-go/blob/master/internal/transport/controlbuf.go#L644) call. So `hc.Conn.Write` will be hang after we made [this](https://github.com/grpc/grpc-go/pull/7371/files#diff-c90a44d2d36c8cc9c90b97b290cce3bdf6a5086bdac1e51e42d0e7b8f415b8b8R2761) write to the conn, which will eventually make the above go-routine to run forever and end up in leaked go-routine.(btw we already have deadline for net.Conn.Write defined [here](https://github.com/grpc/grpc-go/pull/7371/files#diff-5334018444bcb34d15bcb3c8a3050cdf9c8a40fe41305c3a602ea4bfeb0cd178R988) at the top of ct.Close, but then the current test will test only for [this](https://github.com/grpc/grpc-go/pull/7371/files#diff-5334018444bcb34d15bcb3c8a3050cdf9c8a40fe41305c3a602ea4bfeb0cd178R1022) loopyWriter to timeout).
WDYT?

Reply:
Not entirely familiar with which goroutine you are referring to but one suggestion is that you probably don't have to do the actual write `n, err = hc.Conn.Write(b)`? You can just compare the passed bytes `b` if its goaway (like using `string(b)`) and then just hang? That way you don't have any risk of spawning any other go routine (if any).

Reply:
Actually we don't even need to compare it, we can just make sure hangingConn hangs from the time it receives the goAway request. 

@dfawley i have removed the timer logic, and updated the PR.

Reply:
probably good to release the `<-hc.hangConn` at the end of test?